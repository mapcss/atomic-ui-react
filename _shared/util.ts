// This module is browser compatible.

import { isLength0, not } from "../deps.ts";
import { HasFocusElement } from "./types.ts";
export type TempIdReturnValue = {
  readonly current: number;
  readonly next: number;
};
export function tempId(initialState = 0): TempIdReturnValue {
  return {
    get current(): number {
      return initialState;
    },
    get next(): number {
      const _id = initialState;
      initialState++;
      return _id;
    },
  };
}
export const DEFAULT_TEMP_ID: TempIdReturnValue = {
  current: 0,
  next: 0,
};

export function getNextIndex(
  currentIndex: number,
  matrix: Readonly<boolean[]>,
): number {
  if (isLength0(matrix) || matrix.every(not(Boolean))) return currentIndex;
  const _next = currentIndex + 1;
  if (matrix[_next]) return _next;
  if (matrix[_next] === false) {
    return getNextIndex(_next, matrix);
  }

  // if overflow from array, check from start.
  return getNextIndex(-1, matrix);
}

export function getPrevIndex(
  currentIndex: number,
  matrix: Readonly<boolean[]>,
): number {
  if (isLength0(matrix) || matrix.every(not(Boolean))) return currentIndex;
  const _prev = currentIndex - 1;
  if (matrix[_prev]) return _prev;
  if (matrix[_prev] === false) {
    return getPrevIndex(_prev, matrix);
  }
  return getPrevIndex(matrix.length, matrix);
}

export function getFirstIndex(
  currentIndex: number,
  matrix: Readonly<boolean[]>,
): number {
  const index = matrix.findIndex(Boolean);
  if (index < 0) {
    return currentIndex;
  } else {
    return index;
  }
}

export function getLastIndex(
  currentIndex: number,
  matrix: Readonly<boolean[]>,
): number {
  // findLastIndex is ES2022 feature. so Not used for a while.
  // const index = matrix.findLastIndex(Boolean);
  const index = matrix.reduceRight((acc, cur, i) => {
    if (0 <= acc) return acc;
    return cur ? i : acc;
  }, -1);
  if (index < 0) {
    return currentIndex;
  } else {
    return index;
  }
}

const selector =
  'a[href], button, input, textarea, select, details, [tabindex]:not([tabindex="-1"])';

export function filterFocusable(
  root: ParentNode | null | undefined,
): HasFocusElement[] {
  if (!root) return [];
  const els = [...root.querySelectorAll(selector)];

  return els.filter(hasNotInvalidAttribute).filter(hasFocusElement);
}

function hasFocusElement(el: Element): el is HasFocusElement {
  return el instanceof HTMLElement || el instanceof SVGElement ||
    el instanceof MathMLElement;
}

function hasNotInvalidAttribute(el: Element): boolean {
  return !el.hasAttribute("disabled") && !el.getAttribute("aria-hidden");
}
