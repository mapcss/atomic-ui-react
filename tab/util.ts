export function getNextIndex(currentIndex: number, sumCount: number): number {
  const _next = currentIndex + 1;

  if (_next < sumCount) {
    return _next;
  }
  return 0;
}

export function getPrevIndex(currentIndex: number, sumCount: number): number {
  const _prev = currentIndex - 1;
  if (0 <= _prev) {
    return _prev;
  }
  return sumCount - 1;
}

export function getFirstIndex(): number {
  return 0;
}

export function getLastIndex(_: number, sumCount: number): number {
  return sumCount - 1;
}
