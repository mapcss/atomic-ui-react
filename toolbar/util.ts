// This module is browser compatible.

export function calcTabIndex(
  { isActive, hasActivated, isFirst }: {
    isActive: boolean;
    hasActivated: boolean;
    isFirst: boolean;
  },
): number {
  if (isActive || (!hasActivated && isFirst)) return 0;
  return -1;
}
