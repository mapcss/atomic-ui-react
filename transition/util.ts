// This module is browser compatible.

import {
  END,
  ENTER,
  ENTER_FROM,
  ENTER_TO,
  ENTERED,
  INIT,
  LEAVE,
  LEAVE_FROM,
  LEAVE_TO,
  LEAVED,
  START,
  WAIT,
} from "./constant.ts";
import { ReturnValue } from "./use_transition.ts";
import { TransitionLifecycleMap } from "./use_transition_lifecycle.ts";

const ENTER_TRANSITION_MAP: TransitionLifecycleMap = {
  [INIT]: [ENTER_FROM],
  [START]: [ENTER_FROM, ENTER],
  [WAIT]: [ENTER, ENTER_TO],
  [END]: [ENTERED],
};

const LEAVE_TRANSITION_MAP: TransitionLifecycleMap = {
  [INIT]: [LEAVE_FROM],
  [START]: [LEAVE_FROM, LEAVE],
  [WAIT]: [LEAVE, LEAVE_TO],
  [END]: [LEAVED],
};

/** Compute transition duration from `CSSStyleDeclaration` */
export function getDuration(el: Element): number {
  try {
    const { transitionDuration, transitionDelay } = globalThis.getComputedStyle(
      el,
    );

    const duration = parseFiniteFloat(transitionDuration);

    if (!duration) {
      return 0;
    }

    const delay = parseFiniteFloat(transitionDelay) ?? 0;
    const extension = 1000 * (duration + delay);

    return Number.isFinite(extension) ? extension : 0;
  } catch {
    return 0;
  }
}

function parseFiniteFloat(value: string): number | undefined {
  const num = Number.parseFloat(value);
  return Number.isFinite(num) ? num : undefined;
}

export function getTransitionMap(isEnter: boolean): TransitionLifecycleMap {
  return isEnter ? ENTER_TRANSITION_MAP : LEAVE_TRANSITION_MAP;
}

/** It takes two `boolean` values and determines if it can be rendered to the DOM.
 * This can be used to determine rendering of components with transitions applied.
 * @param isShow - Whether visible or not.
 * @param isCompleted - Whether the transition is completed or not.
 *
 * ```ts
 * import { isShowable } from "https://deno.land/x/atomic_ui_react@$VERSION/mod.ts"
 * isShowable(true, {isActivated: false, isCompleted: false })
 * ```
 */
export function isShowable(
  isShow: boolean,
  { isActivated, isCompleted }: Pick<
    ReturnValue,
    "isActivated" | "isCompleted"
  >,
): boolean {
  if (!isActivated) return isShow;
  return isShow || !isCompleted;
}
