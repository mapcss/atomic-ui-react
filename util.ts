// This module is browser compatible.
// deno-lint-ignore-file no-explicit-any

import { isFunction, isObject } from "./deps.ts";
import { Attributes, ReactElement, RefAttributes, RefObject } from "react";

export type Lazyable<T, U extends (...args: any[]) => T = () => T> = T | U;

/** Whether this environment is Browser or not.
 * This is compatible with Node.js
 */
export const isBrowser = !("Deno" in globalThis) && !("process" in globalThis);
export function lazyEval<T, U extends (...args: any[]) => T = () => T>(
  lazyable: Lazyable<T, U>,
): T {
  return isFunction(lazyable) ? lazyable() : lazyable;
}

export function isRefObject<T>(value: unknown): value is RefObject<T> {
  return isObject(value) && "current" in value;
}

export function hasRef<T>(
  element: ReactElement,
): element is ReactElement & RefAttributes<T> {
  return "ref" in element;
}

export function hasRefObject<T>(
  element: ReactElement & RefAttributes<T>,
): element is ReactElement & Attributes & { ref?: RefObject<T> } {
  return isObject(element.ref);
}
