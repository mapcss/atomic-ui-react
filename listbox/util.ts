import { joinChars } from "../util.ts";

export function getIdPrefix(id: string): string {
  return joinChars([id, "listbox", "option"], "-")!;
}
