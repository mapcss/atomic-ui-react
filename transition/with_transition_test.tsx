import { renderToString } from "react-dom/server";
import { render } from "@testing-library/react";
import WithTransition from "./with_transition.ts";
import {
  assertSnapshot,
  describe,
  expect,
  FakeTime,
  it,
  mockGetComputedStyle,
  ParamReturn,
  setupJSDOM,
  setupRaf,
} from "../dev_deps.ts";
import { mergeProps } from "../util.ts";

Deno.test("render as SSR", () => {
  const table: ParamReturn<typeof renderToString>[] = [
    [
      <WithTransition enter="transition" enterFrom="opacity-0" isShow>
        {(attrs) => <div {...attrs}>test</div>}
      </WithTransition>,
      `<div>test</div>`,
    ],
    [
      <WithTransition enter="transition" enterFrom="opacity-0" isShow>
        {(attrs) => <div {...attrs} className="test">test</div>}
      </WithTransition>,
      `<div class="test">test</div>`,
    ],
    [
      <WithTransition
        enter="transition"
        enterFrom="opacity-0"
        isShow
        immediate
      >
        {(attrs) => <div {...attrs}>test</div>}
      </WithTransition>,
      `<div class="opacity-0">test</div>`,
    ],
    [
      <WithTransition
        enter="transition"
        enterFrom="opacity-0"
        isShow
        immediate
      >
        {(attrs) => (
          <div {...mergeProps({ className: "test" }, attrs)}>test</div>
        )}
      </WithTransition>,
      `<div class="test opacity-0">test</div>`,
    ],
    [
      <WithTransition leaveFrom="opacity-0" isShow={false}>
        {(_, { isShowable }) => isShowable ? <div>test</div> : <></>}
      </WithTransition>,
      ``,
    ],
    [
      <WithTransition isShow={false} immediate>
        {(attrs) => (
          <div {...mergeProps({ className: "test" }, attrs)}>test</div>
        )}
      </WithTransition>,
      `<div class="test">test</div>`,
    ],
    [
      <WithTransition leaveFrom="opacity-0" isShow={false} immediate>
        {(attrs) => <div {...attrs}>test</div>}
      </WithTransition>,
      `<div class="opacity-0">test</div>`,
    ],
    [
      <WithTransition leaveFrom="opacity-0" isShow={false} immediate>
        {(attrs) => (
          <div {...mergeProps({ className: "test" }, attrs)}>test</div>
        )}
      </WithTransition>,
      `<div class="test opacity-0">test</div>`,
    ],
    [
      <WithTransition enterFrom="opacity-0" isShow>
        {(attrs) => <div {...attrs}>test</div>}
      </WithTransition>,
      `<div>test</div>`,
    ],
    [
      <WithTransition enterFrom="opacity-0" isShow immediate>
        {(attrs) => <div {...attrs}>test</div>}
      </WithTransition>,
      `<div class="opacity-0">test</div>`,
    ],
    [
      <WithTransition leaveFrom="opacity-80" isShow={false}>
        {(attrs) => <div {...attrs}>test</div>}
      </WithTransition>,
      `<div>test</div>`,
    ],
    [
      <WithTransition leaveFrom="opacity-80" isShow={false} immediate>
        {(attrs) => <div {...attrs}>test</div>}
      </WithTransition>,
      `<div class="opacity-80">test</div>`,
    ],
  ];

  table.forEach(([element, result]) =>
    expect(renderToString(element)).toBe(result)
  );
});

const describeTests = describe({
  name: "WithTransition",
  async beforeEach(this: { time: FakeTime; el: HTMLElement; raf: () => void }) {
    await setupJSDOM();

    this.time = new FakeTime();
    this.el = globalThis.document.createElement("div");
    this.raf = setupRaf();
  },
  afterEach() {
    this.raf();
    this.time.restore();
  },
});

it(
  describeTests,
  "if isShow is true at first mount, it should render to DOM",
  async function (t) {
    const { getByTestId } = render(
      <WithTransition isShow enterFrom="opacity-0">
        {(attrs) => (
          <div
            {...mergeProps({ "data-testid": "test", className: "test" }, attrs)}
          >
            test
          </div>
        )}
      </WithTransition>,
    );

    const id = getByTestId("test");
    expect(id).toBeDefined();
    expect(id.className).toBe("test");
    await assertSnapshot(t, id.outerHTML);
  },
);

it(
  describeTests,
  "should be match snapshot",
  async function (t) {
    const { time, el } = this;
    el.style.transitionDuration = "3s";
    const reset = mockGetComputedStyle(() => el.style);
    const transitionProps = {
      enterFrom: "opacity-0",
      enter: "  transition ",
      enterTo: "opacity-100  ",
      entered: " text-red-500 text-red-500",
      leaveFrom: " opacity-80 ",
      leave: " transition duration",
      leaveTo: "opacity-30",
    };
    const { container, rerender } = render(
      <WithTransition
        isShow
        immediate
        {...transitionProps}
      >
        {(attrs) => (
          <div {...mergeProps({ className: "test" }, attrs)}>test</div>
        )}
      </WithTransition>,
    );

    await assertSnapshot(t, container.innerHTML);
    time.next();
    await assertSnapshot(t, container.innerHTML);
    time.next();
    await assertSnapshot(t, container.innerHTML);
    time.next();
    await assertSnapshot(t, container.innerHTML);
    time.runAll();
    await assertSnapshot(t, container.innerHTML);
    rerender(
      <WithTransition
        isShow={false}
        immediate
        {...transitionProps}
      >
        {(attrs) => (
          <div {...mergeProps({ className: "test" }, attrs)}>test</div>
        )}
      </WithTransition>,
    );
    await assertSnapshot(t, container.innerHTML);
    time.next();
    await assertSnapshot(t, container.innerHTML);
    time.next();
    await assertSnapshot(t, container.innerHTML);
    time.next();
    await assertSnapshot(t, container.innerHTML);
    time.runAll();
    await assertSnapshot(t, container.innerHTML);
    reset();
  },
);

it(
  describeTests,
  "should keep DOM when leaved transition map is exists",
  async function (t) {
    const { time, el } = this;
    el.style.transitionDuration = "3s";
    const reset = mockGetComputedStyle(() => el.style);
    const transitionProps = {
      leaveFrom: " opacity-80 ",
      leave: " transition duration",
      leaveTo: "opacity-30",
      leaved: "opacity-0",
    };
    const { container } = render(
      <WithTransition
        isShow={false}
        immediate
        {...transitionProps}
      >
        {(attrs) => (
          <div {...mergeProps({ className: "test" }, attrs)}>test</div>
        )}
      </WithTransition>,
    );

    await assertSnapshot(t, container.innerHTML);
    time.next();
    await assertSnapshot(t, container.innerHTML);
    time.next();
    await assertSnapshot(t, container.innerHTML);
    time.next();
    await assertSnapshot(t, container.innerHTML);
    time.runAll();
    await assertSnapshot(t, container.innerHTML);

    reset();
  },
);

// it(
//   describeTests,
//   "should render DOM as Props",
//   async function (t) {
//     const { time, el } = this;
//     el.style.transitionDuration = "3s";
//     const reset = mockGetComputedStyle(() => el.style);
//     const transitionProps = {
//       enterFrom: "opacity-0",
//       enter: "  transition ",
//       enterTo: "opacity-100  ",
//       entered: " text-red-500 text-red-500",
//       leaveFrom: " opacity-80 ",
//       leave: " transition duration",
//       leaveTo: "opacity-30",
//       leaved: "hidden",
//     };

//     const Child = forwardRef<HTMLDivElement>((_, ref) => {
//       return <div className=" test " ref={ref}>children</div>;
//     });
//     const { container, rerender } = render(
//       <WithTransition
//         isShow
//         immediate
//         {...transitionProps}
//       >
//         <Child />
//       </WithTransition>,
//     );

//     await assertSnapshot(t, container.innerHTML);
//     time.next();
//     await assertSnapshot(t, container.innerHTML);
//     time.next();
//     await assertSnapshot(t, container.innerHTML);
//     time.next();
//     await assertSnapshot(t, container.innerHTML);
//     time.runAll();
//     await assertSnapshot(t, container.innerHTML);

//     rerender(
//       <WithTransition
//         isShow={false}
//         immediate
//         {...transitionProps}
//       >
//         <Child />
//       </WithTransition>,
//     );
//     await assertSnapshot(t, container.innerHTML);
//     time.next();
//     await assertSnapshot(t, container.innerHTML);
//     time.next();
//     await assertSnapshot(t, container.innerHTML);
//     time.next();
//     await assertSnapshot(t, container.innerHTML);
//     time.runAll();
//     await assertSnapshot(t, container.innerHTML);

//     reset();
//   },
// );
