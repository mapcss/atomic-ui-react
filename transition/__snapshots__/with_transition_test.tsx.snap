export const snapshot = {};

snapshot[`WithTransition > if isShow is true at first mount, it should render to DOM 1`] = `'<div data-testid="test" class="test">test</div>'`;

snapshot[`WithTransition > should be match snapshot 1`] = `'<div class="test opacity-0">test</div>'`;

snapshot[`WithTransition > should be match snapshot 2`] = `'<div class="test opacity-0 transition">test</div>'`;

snapshot[`WithTransition > should be match snapshot 3`] = `'<div class="test transition opacity-100">test</div>'`;

snapshot[`WithTransition > should be match snapshot 4`] = `'<div class="test text-red-500">test</div>'`;

snapshot[`WithTransition > should be match snapshot 5`] = `'<div class="test text-red-500">test</div>'`;

snapshot[`WithTransition > should be match snapshot 6`] = `'<div class="test opacity-80">test</div>'`;

snapshot[`WithTransition > should be match snapshot 7`] = `'<div class="test opacity-80 transition duration">test</div>'`;

snapshot[`WithTransition > should be match snapshot 8`] = `'<div class="test transition duration opacity-30">test</div>'`;

snapshot[`WithTransition > should be match snapshot 9`] = `'<div class="test">test</div>'`;

snapshot[`WithTransition > should be match snapshot 10`] = `'<div class="test">test</div>'`;

snapshot[`WithTransition > should keep DOM when leaved transition map is exists 1`] = `'<div class="test opacity-80">test</div>'`;

snapshot[`WithTransition > should keep DOM when leaved transition map is exists 2`] = `'<div class="test opacity-80 transition duration">test</div>'`;

snapshot[`WithTransition > should keep DOM when leaved transition map is exists 3`] = `'<div class="test transition duration opacity-30">test</div>'`;

snapshot[`WithTransition > should keep DOM when leaved transition map is exists 4`] = `'<div class="test opacity-0">test</div>'`;

snapshot[`WithTransition > should keep DOM when leaved transition map is exists 5`] = `'<div class="test opacity-0">test</div>'`;
